text/0000-reformed-fn-pointers.md

```diff
@@ -0,0 +1,96 @@
+
+1. Function item types: the types of function items, `fn(arg_list) -> ret_type {foo}`s.
+2. Function pointer types: the types of pointers to functions, `fn(arg_list) -> ret_type`s.
+
+Though called "function pointers", the function pointer types are considered *value* types (not reference/pointer types) by the language, which has the following problems:
+
+1. Inconsistencies in the language, especially when it comes to FFI codes that deal with nullable foreign function pointers. 
+2. There is currently no easy way to express the lifetime constraint on a function pointer to a non-`'static` function, which in particular makes it harder to implement type-safe JIT-ed functions/hot-loaded plugins in the future.
+
+Thus, this RFC proposes the following solution.
+
+# Detailed design
+
+Repurpose current function pointer types to denote "incomplete function items". Introduce function reference types of the form `&fn(arg_list) -> ret_type` and new function pointer types of the form `*const fn(arg_list) -> ret_type`.
+
+In the following section, `fn{f}`s, `fn`s, `&fn`s and `*const fn`s denote function item types, current function pointer types, function reference types and new function pointer types, respectively. Those types are considered "compatible" if their `(arg_list) -> ret_type` parts match.
+
+The following rules will apply:
+
+1. `fn{f}`s are still function item types, or "function handles/proxies/values", whose representations and semantics remain unchanged.
+2. `fn`s are no longer function pointer types, but types representing "incomplete function items", which are unsized statically and zero-sized dynamically.
+3. `&fn`s are function reference types, DST pointers with auxiliary data of type `()`, and work like other references.
+4. `*const fn`s are raw function pointer types that work as normally expected, and casting between compatible `&fn`s and `*const fn`s is valid.
+5. There are no `&mut fn`s, also no `*mut fn`s.
+6. `fn{f}`s still implement the closure traits (`Fn`/`FnMut`/`FnOnce`).
+7. `fn`s still implement the closure traits, for keeping `&fn`s coercible to closure trait objects.
+8. `&fn`s will implement the closure traits.
+9. The `fn{f} -> fn` coercions between compatible `fn{f}`s and `fn`s are no longer valid.
+10. The `&fn{f} -> &fn` coercions between compatible `fn{f}`s and `fn`s are valid as unsizing coercions.
+
+Optional changes that can be applied now or after Rust's final stabilization:
+
+1. Make `fn{f}`s zero-sized statically to save space and better align with the fact that `fn`s are zero-sized dynamically.
+2. Make `&fn{f}`s implement closure traits for better symmetry between `fn{f}`s and `fn`s. 
+
+Notes:
+
+1. Currently, both `&fn{f}`s and `&fn`s are coercible to closure trait objects, but are not closures themselves. After the changes, they will be closures (`&fn`s) or coercible to closures (`&fn{f}`s). If the second optional change happens, then `&fn{f}`s will also be closures, without coercions.
```

Comment:
Eh, I said

> both `&fn{f}`s and `&fn`s are coercible to closure trait objects,

I think the `&` before `fn{f}` and `fn` makes this statement true? I was not saying `fn{f}`s or `fn`s are coercible to closure trait objects.